### ISSUE TITLE
unable to access kubeflow-dashboard from outside the cluster

### ISSUE THREAD

**Issue by tomalbrecht** [kubeflow/manifests#974]
## Steps to reproduce

Setup kubeflow on microk8s as described here: https://www.kubeflow.org/docs/started/k8s/kfctl-istio-dex/ (and adjust istio configuration as described here:  https://github.com/kubeflow/manifests/issues/959)

When all pods came up, use port-forwarding to check the kubeflow dashboard as described here: https://www.kubeflow.org/docs/started/k8s/kfctl-istio-dex/#accessing-kubeflow

```
kubectl port-forward -n istio-system service/istio-ingressgateway 8081:443 --address=0.0.0.0
kubectl port-forward -n istio-system service/istio-ingressgateway 8081:80 --address=0.0.0.0 # does not work either
```

When accessing the host from another computer (client) on `https://hostnameORip:8081`, the following error will show up.
```
E0303 08:52:14.438417   32128 portforward.go:400] an error occurred forwarding 8081 -> 443: error forwarding port 443 to pod 0fa60992b404d06a4fb09cf240dc53297316da8f2daddc0fc3ab1f602248238b, uid : failed to execute portforward in network namespace "/var/run/netns/cni-fc82cc55-60f1-9cdd-09ac-f07086cc0483": socat command returns error: exit status 1, stderr: "2020/03/03 08:52:14 socat[2807] E connect(5, AF=2 127.0.0.1:443, 16): Connection refused\n"
```

## Loadbalancer

Using MetalLB as described [here](https://www.kubeflow.org/docs/started/k8s/kfctl-istio-dex/#expose-with-a-loadbalancer) results in the same blank dashboard page. I'll provide the log files of the pod as soon as I got them.

## Portforwarding centraldashboard

centraldashboard is running. Forwarding the service will show the board. So it seems a problem routing the dashboard over istio-ingressgateway.

`kubectl port-forward -n kubeflow service/centraldashboard 8081:80 --address=0.0.0.0`

## Setup
I used the following configuration for installation:
microk8s: 1.14/stable + https://github.com/kubeflow/manifests/issues/959
CONFIG_URI: https://raw.githubusercontent.com/kubeflow/manifests/v1.0-branch/kfdef/kfctl_istio_dex.v1.0.0.yaml
kfctl: kfctl v1.0-rc.3-1-g24b60e8 from https://github.com/kubeflow/kfctl/releases/download/v1.0/kfctl_v1.0-0-g94c35cf_linux.tar.gz
OS: ubuntu 18.04 lts

I already googled and could not find a matching solution to my problem. Any hint?

## References to other Repositories and Issues
These Issues seem to have similar problems or at least have something in common.
https://github.com/kubeflow/kubeflow/issues/4813
https://github.com/kubeflow/kubeflow/issues/4819

## Logfiles
[istio-ingressgateway-68875ff4cd-gkvmr.log](https://github.com/kubeflow/manifests/files/4286246/istio-ingressgateway-68875ff4cd-gkvmr.log)
[metadata-envoy-deployment-5688989bd6-gkt2h.log](https://github.com/kubeflow/manifests/files/4286383/metadata-envoy-deployment-5688989bd6-gkt2h.log)
[centraldashboard-869f55ddbd-r8ztf.log](https://github.com/kubeflow/manifests/files/4286400/centraldashboard-869f55ddbd-r8ztf.log)

## Additional Tests

I setup a nginx service to eliminate network and routing problems between the two machines. nginix works as expected (locally and externally). 
```
kubectl create deployment nginx --image=nginx
kubectl create service nodeport nginx --tcp=80:80
```
When setting up the cluster with microk8s v1.15 and Kubeflow 1.0.1 everything works as expected. I am able to connect to the ingress-gateway via Nodeport (Port 31380 for http and 31390 for https).

As soon as I reboot the Server, this way will break. I won't be able to connect via NodePort anymore, even the Service still seems to be a NodePort Service.

**Comment by swiftdiaries** [comment]
I think the `443` port is https traffic. The envoy filter for Dex is enabled for the http port at port 80. 
Could you try port-forwarding to the istio-ingressgateway service at port 80?

**Comment by tomalbrecht** [comment]
`kubectl port-forward -n istio-system service/istio-ingressgateway 8081:80 --address=0.0.0.0` does not work either. Same result. It takes around 5 seconds, then I get a blank page.

<img width="298" alt="Bildschirmfoto 2020-03-09 um 12 27 31" src="https://user-images.githubusercontent.com/17781570/76209133-6956b180-6201-11ea-82e5-8484cb5b55f7.png">

**Comment by tomalbrecht** [comment]
Port seems to be reachable from the outside.
<img width="410" alt="image" src="https://user-images.githubusercontent.com/17781570/76291229-fb67c400-62ac-11ea-9760-ed797d56e75c.png">

**Comment by tomalbrecht** [comment]
I setup an nginx to eliminate network and routing problems. nginix works as expected. 
```
kubectl create deployment nginx --image=nginx
kubectl create service nodeport nginx --tcp=80:80
```
Right now for me it seems like the ingress-gateway is not working as expected.

**Comment by tomalbrecht** [comment]
- Same behavior for v1.0.1.
- NodePort ist not accessible either

**Comment by tomalbrecht** [comment]
Some additional information. When setting up the cluster with  microk8s v1.15 and Kubeflow 1.0.1 everything works as expected. I am able to connect to the ingress-gateway via Nodeport (Port 31380 for http and 31390 for https).

As soon as I reboot the Server, this way will break. I won't be able to connect via NodePort anymore, even the Service still seems to be a NodePort Service.

**Comment by tomalbrecht** [comment]
We were able to setup Kubeflow v1.0.1 with microk8s v1.15 and LoadBalancer. The Dashboard is available and working. Thus I am closing this issue.

We had no luck installing Kubeflow v1.0.0 nor v1.0.1 on microk8s v1.14.

Maybe the documentation could be updated. There are kubectl commands which should only work in kubernetes 1.15+.

**Comment by TomLaMantia** [comment]
> 
> 
> We were able to setup Kubeflow v1.0.1 with microk8s v1.15 and LoadBalancer. The Dashboard is available and working. Thus I am closing this issue.
> 
> We had no luck exposing the Dashboard with LoadBalancer (metallb), Kubeflow v1.0.0 nor v1.0.1 on microk8s v1.14.
> 
> Maybe the documentation could be updated. There are kubectl commands which should only work in kubernetes 1.15+.

I am having very similar issues, would you please share some more details? Thanks

**Comment by tomalbrecht** [comment]
The documentation has been updated by the team. But you need to know which points are mandatory and there are still some pitfalls. 

Use this doc: https://www.kubeflow.org/docs/started/k8s/kfctl-istio-dex/

important: wait for all pods being up (can take 30 to 45 minutes, depending on your machine)

- install microk8s (version 1.15/stable)
- update Istio configuration, as described [here](https://github.com/kubeflow/manifests/issues/959) (or see below)
- wait until all pods in your cluster are ready
- follow description at "setup and deploy kubeflow"
- follow description at "Expose Kubeflow" (important: setup SSL, even if you don't want to use it)
- follow description at "Expose with a Loadbalancer" (important: install MetalLB, you have to click on the triangle for more information on this)

for microk8s:
```
# kubernetes / Istio configuration for trustworthy JWTs
# https://github.com/kubeflow/manifests/issues/959
# check if entries already exist, otherwise add them
grep -qxF "service-account-signing-key-file" /var/snap/microk8s/current/args/kube-apiserver \
|| echo "--service-account-signing-key-file=\${SNAP_DATA}/certs/serviceaccount.key" \
>> /var/snap/microk8s/current/args/kube-apiserver
grep -qxF "service-account-issuer" /var/snap/microk8s/current/args/kube-apiserver \
|| echo "--service-account-issuer=kubernetes.default.svc" \
>> /var/snap/microk8s/current/args/kube-apiserver

```

**Comment by paravatha** [comment]
@tomalbrecht 
Thanks a lot for finding the issues and providing suggestions. 
After running these port-forward commands, I am able to to access Dashboard, Katib and Pipleine components on different ports

```
kubectl port-forward -n kubeflow service/kubeflow-dashboard 8082:8082 --address=0.0.0.0 &
kubectl port-forward -n kubeflow service/pipelines-ui 8086:3000 --address=0.0.0.0 &
kubectl port-forward -n kubeflow service/katib-ui 8084:8000 --address=0.0.0.0 &
```
I am still not able to access each UI component directly from kubeflow dashboard
It gives 
"Sorry, /pipeline/ is not a valid page"
"Sorry, /katib/ is not a valid page"
"Sorry, /metadata/ is not a valid page"

Any idea what I am missing?

**Comment by tomalbrecht** [comment]
You have to login first (via dex), so the cookies are set. The istio gateway will forward you then to the main dashboard. From there you can access all other pages. So the endpoints (as far as I understand) should not be accessible without authentication.

`kubectl get services -n istio-system | grep gateway` will show you the gateway service (when you did microk8s enable kubeflow then its called "ambassador")

```bash
istio-ingressgateway     LoadBalancer   10.152.183.3     10.0.0.100    15020:31937/TCP,80:31380/TCP,443:31390/TCP,31400:31400/TCP,15029:32294/TCP,15030:30113/TCP,15031:30194/TCP,15032:31693/TCP,15443:31912/TCP
```

The port you are looking for is 31390 (for ssl). 31380 did not work for me. When you setup the load balancer (metallb) you will see 'Loadbalancer' as service type. I only got it working this way. NodePorts + Kubeflow + microk8s did not work.

Thus you should be able to connect to your gateway using your external IP (not the 10.0.0.100 address):
```bash
echo "Kubeflow Login: https://$(echo $(hostname -I) | tr -s ' ' | cut -d ' ' -f 1):31390"
Kubeflow Login: https://192.168.61.86:31390
```

**Comment by Dimiftb** [comment]
Hi @tomalbrecht and everyone else,

I have a very similar issue to the one described. I've followed the same guide to deploy Kubeflow on a local Kubernetes cluster created with kubeadm. 

I needed a more permanent solution than port forwarding so I've set up metal load balancer and tested with dummy nginx server as described in guide https://www.kubeflow.org/docs/started/k8s/kfctl-istio-dex/#expose-with-a-loadbalancer. 

I've also created a certificate with cert managed as described in the guide. 

I attempted to access the dashboard both with http port 30795 and https 31390 on Istio-ingressgateway service `80:30795/TCP, 443:30934/TCP` and none of them are working. I also tested with Nmap to see if either of the ports are open, but no luck.

The only thing different from @tomalbrecht's setup is that I'm not using dex as this is still in development, so I just want to connect anyhow (with static user) before I setup authentication.

I'm not sure what to do. Could anyone provide any advice or ideas how to debug this? I can supply any logs, or specs, yaml, anything that could help identify what the issue might be.

**Comment by tomalbrecht** [comment]
Are you are of this installation? It's much easier: https://www.kubeflow.org/docs/started/workstation/kubeflow-on-microk8s/

But in case you need a deployment with manifests, I assume you are using these manifests, right https://raw.githubusercontent.com/kubeflow/manifests/v1.2-branch/kfdef/kfctl_k8s_istio.v1.2.0.yaml.

- When you setup the loadbalancer, setting the right IP and Hostmame is essential.
- The order of steps is also essential. 
    1. You have deploy metallb
    2. patch istio-gateway to use it
    3. wait until you get `.status.loadBalancer.ingress[0]``
    4. create the certificate
    5. acces the cluster as you described (the ports you described seem to be in the correct range). Sometimes you have to wait 1-2 minutes until k8 set everything up
- I would use another clean setup to repeat these steps in case they did not work. If you missed one step, it seems easier for me to clean the environment 
- We had trouble with these self signed certificates. You can disable this, when you disable `httpsRedirect` setting it to `false`
- a simpler way (instead of loadbalancer) could be Nodeports. Instead of setting up a Loadbalancer, the only thing you have to do is, patching ingres-gateway to NodePort instead of LoadBalancer

I hope one of these hints will work for you. Good luck!

**Comment by Dimiftb** [comment]
Hi @tomalbrecht, 

I thought I'd write back to share how my situation resolved in case it helps someone else. 

In the end it turned out that the loadBalancer was working correctly and the dashboard and the nginx test service were accessible when I was connected to the network of my company. However, I still couldn't access them from my own machine at home.
I managed to isolate the issue to being unable to access the dashboard only when I'm not connected to my company's network. 
It turned out that they had a firewall blocking all connections coming from outside the network and I had to use a VPN. Once I connected through the VPN everything worked as expected. 

Thank you @tomalbrecht and everyone for your help. Hopefully this helps someone else.
